sql/sql_repl.cc

```diff
@@ -2232,15 +2250,130 @@ send_event_to_slave(binlog_send_info *info, Log_event_type event_type,
   }
 
   /*
-    Skip events with the @@skip_replication flag set, if slave requested
-    skipping of such events.
+    Skip events with the @@skip_replication flag set, i.e. either if slave 
+    requested skipping of such events, or if filtered via the binlog_dump_* 
+    filter options
   */
-  if (info->thd->variables.option_bits & OPTION_SKIP_REPLICATION)
+  if (info->thd->variables.option_bits & OPTION_SKIP_REPLICATION ||
+    !binlog_dump_filter->is_db_empty() || binlog_dump_filter->is_on())
   {
+
     uint16 event_flags= uint2korr(&((*packet)[FLAGS_OFFSET + ev_offset]));
 
-    if (event_flags & LOG_EVENT_SKIP_REPLICATION_F)
-      return NULL;
+    const uchar *buf= (const uchar*) packet->ptr() + ev_offset;
+
+      Log_event_type etype= (Log_event_type) buf[EVENT_TYPE_OFFSET];
+
+      // Only ROW events have m_flags
+      bool is_rows_event =
+          (etype == WRITE_ROWS_EVENT)        ||
+          (etype == UPDATE_ROWS_EVENT)       ||
+          (etype == DELETE_ROWS_EVENT)       ||
+          (etype == WRITE_ROWS_EVENT_V1)     ||
+          (etype == UPDATE_ROWS_EVENT_V1)    ||
+          (etype == DELETE_ROWS_EVENT_V1);
+      bool is_gtid_event= (etype == GTID_EVENT);
+
+    if (is_gtid_event)
+    {
+      info->sent_unfiltered_row_event= false; // reset per event group
+    }
+    
+    if ((event_flags & LOG_EVENT_SKIP_REPLICATION_F))
+    {
+      /*
+        if the current event has the STMT_F flag as well as the
+        LOG_EVENT_SKIP_REPLICATION_F flag then we should clear the buffer
+        data of this event and rewrite this *_ROW_EVENT to avoid sending
+        filtered data to the replica otherwise if it does not have
+        the STMT_F flag then it should just get filterd 
+        (return NULL basically filters and does not send the event
+        over the network)
+      */
+
+      if (!is_rows_event) {
+          return NULL;
+      }
+
+      /*
+        Extract the STMT_END_F of the rows_event flag
+      */
+      uint8 common_header_len= info->fdev->common_header_len;
+
+      uint8 post_header_len= info->fdev->post_header_len[etype-1];
+
+      const uchar *post_start= buf + common_header_len;
+
+      post_start+= RW_MAPID_OFFSET;
+
+      uint64 table_id;
+
+      if (post_header_len == 6)
+      {
+          table_id= uint4korr(post_start);
+          post_start+= 4;
+      }
+      else
+      {
+        table_id= uint4korr(post_start);
+        table_id|= ((uint64) uint2korr(post_start + 4)) << 32;
+        // skip 6 bytes
+        post_start+= RW_FLAGS_OFFSET;
+      }
+
+      uint16 m_flags= uint2korr(post_start);
+
+      bool has_stmt_end_f= (m_flags & Rows_log_event::STMT_END_F);
+
+
+      if (!has_stmt_end_f)
```

Comment:
For easy comparisons like this, it is easier for others to read to just have the comparison:

```suggestion
if (!(m_flags & Rows_log_event::STMT_END_F))
```

Then there is also weird spacing above